torchdata/dataloader2/communication/iter.py

```diff
@@ -208,3 +213,62 @@ def nonblocking_next(self):
         if isinstance(response, communication.messages.InvalidStateResponse):
             raise NotAvailable
         return response.value
+
+
+class _IterateQueueDataPipes(IterDataPipe):
+    r"""
+    Takes in ``QueueWrapper``s and iterates through them in a round-robin manner to get batches one-by-one.
+
+    Typically, each worker has one ``QueueWrapper``.
+    """
+
+    def __init__(self, datapipes):
+        # TODO(VitalyFedyunin): Consider combining _IterateQueueDataPipes and QueueWrapper
+        # into one class, which supports any number of queues.
+        self.datapipes = datapipes
+        for dp in self.datapipes:
+            if not isinstance(dp, QueueWrapper):
+                raise Exception("Source datapipes should be an instance of iter.QueueWrapper")
+
+    def __iter__(self):
+        total_pipes = len(self.datapipes)
+        disabled_pipe = [False] * len(self.datapipes)
+        cnt_disabled_pipes = 0
+
+        for idx in range(total_pipes):
+            self.datapipes[idx].protocol.request_next()
+
+        while cnt_disabled_pipes < total_pipes:
+            for idx in range(total_pipes):
+                if not disabled_pipe[idx]:
+                    response = self.datapipes[idx].protocol.get_response_next(block=True)
+                    if isinstance(response, communication.messages.StopIterationResponse):
+                        disabled_pipe[idx] = True
+                        cnt_disabled_pipes += 1
+                        continue
+                    if isinstance(response, communication.messages.InvalidStateResponse):
+                        raise communication.iter.InvalidStateResetRequired
+                    if isinstance(response, communication.messages.TerminateResponse):
+                        raise communication.iter.TerminateRequired
+                    self.datapipes[idx].protocol.request_next()
+                    yield response.value
+
+    def reset(self):
+        # NonBlocking DataPipes do not reset automatically, have to do it manually
```

Comment:
@ejguan Just noticed that the `reset` method has changed after the move. It used to have this:
```python
# Collect all existing requests results to clear queues
        for dp in self.datapipes:
            if dp.protocol.waiting_for_response():
                dp.protocol.get_response_next(block=True)
```

Is this no longer necessary?

Reply:
I don't think it's necessary because we always want to do `reset_epoch` as well for `NonBlocking`, which will discard all existing requests. So, when at the point of `reset`, we should expect no request within the worker process queues.